### PR TITLE
Experiment with Early Hints

### PR DESCRIPTION
**Pull request by yschimke**
For discussion, not connected between Http2Connection/Stream that receives the EarlyHints headers, and Exchange/EventListener for reporting them.

Early Hints should be delivered before the Response headers are available, and the early hints may not match actual headers, so we can't delay and merge them to keep same API.  So using the EventListener seems like the right place to use this.

The challenge I have is that normal flow is driven in CallServerInterceptor, and Exchange is the place where we have the Http2ExchangeCodec and EventListener.  Not clear how to pass

Other questions: For the full set of 100 responses https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/100, is there anything else to consider?

100 Continue - implemented in CallServerInterceptor
101 Switching Protocols - not implemented, we use ALPN, be good to add https://github.com/square/okhttp/issues/5874
103 - This PR.

### CODE REVIEW

File: okhttp/src/jvmTest/java/okhttp3/EarlyHintsTest.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  val eventListener = RecordingEventListener()
+
+  private val client = clientTestRule.newClientBuilder()
+    .eventListenerFactory(clientTestRule.wrap(eventListener))
+    .build()
+
+  @Test
+  fun testCloudflare() {
+    val request = Request.Builder().url("https://tradingstrategy.ai")
+      .header("User-Agent", "curl/7.77.0")
```
Comment (yschimke) [inline okhttp/src/jvmTest/java/okhttp3/EarlyHintsTest.kt:38]: Required to allow Early Hints. I guess User Agent filtering is making sure OkHttp doesn't get broken?

**Comment by yschimke** [comment]
Should HTTP/1.1 early hints be fully supported, or just tested to confirm it's not breaking?

**Comment by gesellix** [comment]
For #5874, support for "101 Switching Protocols" for HTTP/1 would be great.

**Comment by yschimke** [comment]
@gesellix Yep - but not sure it's incremental enough to just throw it in with this. Maybe it is and we can handle these cleanly in a common approach, but switching from HTTP/1.1 to HTTP/2 seems a lot more involved.

**Comment by gesellix** [comment]
@yschimke yes, I certainly didn't want to put any pressure on this. See my comment only as reaction on your thoughts about "is there anything else to consider?" and me trying to keep the awareness about the other issue. Status 103 isn't the same as 101, so this is fine with me :)

**Comment by yschimke** [comment]
Its a good time to look to complete this area.

**Comment by swankjesse** [comment]
An alternative: use the existing EventListener.responseHeaders() function for this, but also for all 1xx responses.

That means users also receive other 1xx codes for free, and the Response type of that API means they get the status code.

**Comment by yschimke** [comment]
Should I just handle it like the continue logic, read headers multiple times until not a 1xx

**Comment by yschimke** [comment]
And create a Response to pass to the listener?

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/http/CallServerInterceptor.kt [hunk 2]
```diff
@@ -103,8 +103,10 @@ class CallServerInterceptor(private val forWebSocket: Boolean) : Interceptor {
           .receivedResponseAtMillis(System.currentTimeMillis())
           .build()
       var code = response.code
-      if (code == 100) {
-        // Server sent a 100-continue even though we did not request one. Try again to read the
+      while (code in 100..199) {
+        exchange.eventListener.responseHeadersEnd(exchange.call, response)
+
+        // Server sent a 1xx so report and continue. Try again to read the
```
Comment (yschimke) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/http/CallServerInterceptor.kt:109]: Seems like I shouldn't ignore a "Switching Protocols" here. 🤔

Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/http/CallServerInterceptor.kt:109]: I think it's a future-proof default. What does curl do with HTTP 105??

Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/http/CallServerInterceptor.kt:109]: I assume this is broken for 101?

**Comment by swankjesse** [comment]
I think this might benefit from a bigger uglier loop to handle combinations of 103 and 100? Seems possible that different middlewares might give us a 103 followed by a 100, or vice versa.

**Comment by yschimke** [comment]
> I think this might benefit from a bigger uglier loop to handle combinations of 103 and 100? Seems possible that different middlewares might give us a 103 followed by a 100, or vice versa.

I think this leaves us with 2 options, and implementing 1 or 2.

a) A minimal change to put tight loops around reading the response headers, and looping with 103 and spitting out to the listener.
b) A bigger change - to cleanly support 1xx overall. I think refactoring the bigger loop is part of this, and it should include Switching Protocols rather than try to get a cleaner change before we consider this path.

Should we bother with a) first? Or aim just for b?)?

**Comment by yschimke** [comment]
> An alternative: use the existing EventListener.responseHeaders() function for this, but also for all 1xx responses.
> 
> That means users also receive other 1xx codes for free, and the Response type of that API means they get the status code.

This feels like a breaking change, which is ok for 5.0, because clients need to know to ignore responseBody sometimes as it may not relate to the response they are waiting for.

**Comment by swankjesse** [comment]
I think `b` is the place to go first. Also we might need some help from MockWebServer here. Maybe this?

```
class MockResponse {
  ...
  val informationalResponses = List<MockResponse>
}
```

**Comment by yschimke** [comment]
Yep, makes sense.

**Pull request closed without merging**